pkg/nvml/device.go

```diff
@@ -15,9 +15,42 @@
 package nvml
 
 import (
+	"fmt"
+	"reflect"
 	"unsafe"
 )
 
+// nvmlDeviceHandle attempts to convert a device d to an nvmlDevice.
+// This is required for functions such as GetTopologyCommonAncestor which
+// accept Device arguments that need to be passed to internal nvml* functions
+// as nvmlDevice parameters.
+func nvmlDeviceHandle(d Device) nvmlDevice {
+	if nvmlDevice, ok := d.(nvmlDevice); ok {
+		return nvmlDevice
+	}
+
+	val := reflect.ValueOf(d)
+	if val.Kind() == reflect.Ptr {
```

Comment:
I had used a for loop here to cover cases where we have a pointer to a pointer?
```suggestion
	for val.Kind() == reflect.Ptr {
```

Reply:
OK. That's fair. Could you also provide an example for the unit test?

Reply:
I can't seem to come up with one very easily. It may even be that since the interface is defined on `nvmlDevice` and not `*nvmlDevice` that it's actually impossible to construct a use-case where this would be needed. Feel free to ignore it if you don' think it's necessary. It likely won't come up in practice.

Reply:
However, this legitimate/real use-case breaks the current implementation:
```
diff --git a/pkg/nvml/device_test.go b/pkg/nvml/device_test.go
index 9dd4c07..a4b9ea3 100644
--- a/pkg/nvml/device_test.go
+++ b/pkg/nvml/device_test.go
@@ -27,6 +27,10 @@ func TestGetTopologyCommonAncestor(t *testing.T) {
        Device
    }

+   type wrappedDevice2 struct {
+       wrappedDevice
+   }
+
    testCases := []struct {
        description string
        device      Device
@@ -43,6 +47,17 @@ func TestGetTopologyCommonAncestor(t *testing.T) {
                Device: nvmlDevice{},
            },
        },
+       {
+           description: "pointer to a pointer of a wrapped device",
+           device: func() Device {
+               d := &wrappedDevice2{
+                   wrappedDevice: wrappedDevice{
+                       Device: nvmlDevice{},
+                   },
+               }
+               return d
+           }(),
+       },
    }

    for _, tc := range testCases {
```

Reply:
```
panic: reflect.Value.Interface: cannot return value obtained from unexported field or method [recovered]
	panic: reflect.Value.Interface: cannot return value obtained from unexported field or method
```

Reply:
This seems to fix that:
```
diff --git a/pkg/nvml/device.go b/pkg/nvml/device.go
index 8daeedc..8a871d5 100644
--- a/pkg/nvml/device.go
+++ b/pkg/nvml/device.go
@@ -43,10 +43,21 @@ func nvmlDeviceHandle(d Device) nvmlDevice {
 		if !typ.Field(i).Anonymous {
 			continue
 		}
-		if _, ok := val.Field(i).Interface().(Device); !ok {
+
+		// Make a copy of the original value to ensure that it (as well as all
+		// of its embedded fields) are addressable.
+		addressableVal := reflect.New(val.Type()).Elem()
+		addressableVal.Set(val)
+
+		// Create a copy of the variable at val.Field(i) that we know we can access
+		// (even if its an unexported type).
+		field := addressableVal.Field(i)
+		field = reflect.NewAt(field.Type(), unsafe.Pointer(field.UnsafeAddr())).Elem()
+
+		if _, ok := field.Interface().(Device); !ok {
 			continue
 		}
-		return nvmlDeviceHandle(val.Field(i).Interface().(Device))
+		return nvmlDeviceHandle(field.Interface().(Device))
 	}
 	panic(fmt.Errorf("unable to convert %T to nvmlDevice", d))
 }
diff --git a/pkg/nvml/device_test.go b/pkg/nvml/device_test.go
index 9dd4c07..bf7f09a 100644
--- a/pkg/nvml/device_test.go
+++ b/pkg/nvml/device_test.go
@@ -27,6 +27,10 @@ func TestGetTopologyCommonAncestor(t *testing.T) {
 		Device
 	}

+	type wrappedDevice2 struct {
+		wrappedDevice
+	}
+
 	testCases := []struct {
 		description string
 		device      Device
@@ -43,6 +47,14 @@ func TestGetTopologyCommonAncestor(t *testing.T) {
 				Device: nvmlDevice{},
 			},
 		},
+		{
+			description: "nested embedded device with concrete type",
+			device: wrappedDevice2{
+				wrappedDevice: wrappedDevice{
+					Device: nvmlDevice{},
+				},
+			},
+		},
 	}

 	for _, tc := range testCases {
```

Reply:
Or maybe this alternative implementation is clearer:
```
func nvmlDeviceHandle(d Device) nvmlDevice {
    var helper func(val reflect.Value) nvmlDevice
    helper = func(val reflect.Value) nvmlDevice {
        if val.Kind() == reflect.Interface {
            val = val.Elem()
        }

        if val.Kind() == reflect.Ptr {
            val = val.Elem()
        }

        if val.Type() == reflect.TypeOf(nvmlDevice{}) {
            return val.Interface().(nvmlDevice)
        }

        if val.Kind() != reflect.Struct {
            panic(fmt.Errorf("unable to convert non-struct type %v to nvmlDevice", val.Kind()))
        }

        for i := 0; i < val.Type().NumField(); i++ {
            if !val.Type().Field(i).Anonymous {
                continue
            }
            if !val.Field(i).Type().Implements(reflect.TypeOf((*Device)(nil)).Elem()) {
                continue
            }
            return helper(val.Field(i))
        }
        panic(fmt.Errorf("unable to convert %T to nvmlDevice", d))
    }
    return helper(reflect.ValueOf(d))
}
```

Reply:
I like the helper-based implementation more -- although it's not clear what the benefits are over just calling the top-level function recursively instead of calling `helper` recursively.

Have updated and added some more tests.

Reply:
The problem is that we can't call `Interface()` on a non-exported field to get its concrete type (which is the case covered by my new test). And without the ability to do that, we can't call the top-level function since it takes a concrete type (as opposed to a `reflect.Value`).